contracts/contract_call.go

```diff
@@ -0,0 +1,116 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+func NewBoundMethod(contractAddress common.Address, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return NewMethod(abi, methodName, maxGas).Bind(contractAddress)
+}
+
+func NewRegisteredContractMethod(registryId common.Hash, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return &BoundMethod{
+		Method: NewMethod(abi, methodName, maxGas),
+		resolveAddress: func(vmRunner vm.EVMRunner) (common.Address, error) {
+			return resolveAddressForCall(vmRunner, registryId, methodName)
+		},
+	}
+}
+
+type BoundMethod struct {
+	Method
+	resolveAddress func(vm.EVMRunner) (common.Address, error)
+}
+
+func (bm *BoundMethod) Query(vmRunner vm.EVMRunner, result interface{}, args ...interface{}) error {
+	return bm.run(vmRunner, result, true, nil, args...)
+}
+
+func (bm *BoundMethod) Execute(vmRunner vm.EVMRunner, result interface{}, value *big.Int, args ...interface{}) error {
```

Comment:
Can we remove the value parameter, its set to 0 in all calls. And in fact the same is true for the EVMRunner. I cant imagine a situation where the system would need to transfer value.

Reply:
I tend to make a different "balance" between the "you ain't gonna need it" and "it's logical to have it".

To me, it's logical to have the `value`, even if we don't use it today. Makes sense for an EVMRunner to be able to define the value on a write call, as it's part of the evm protocol.

It's true that we don't use it today, still. I don't think it adds noise; nor it adds a level of generality that it's "too much".

To be a bit extreme, is like a TDD example of having a `sum(x,y)` function that needs to pass the test `sum(4,5) == 9`. The implementation can be `return 9` and doing `return x + y` is much more than you actually need. Still, doing the latter "knowing" that it makes sense for the `sum` function to do it, it's ok. I know this is very extreme, but just to make a point.

Reply:
The system does value transfers sometimes, currently only CELO transfer, doing `state.SetBalance()` but it would't be unthinkable to do a value transfer in a contract call in the future.

Also, i see that the EVMRunner could prove to be more useful than just the system call. For example for the testing library o mycelo; you can do calls, transfer, etc, operating directly to the EVM/State layer.

Reply:
If its just coincidental that _right now_ we do not happen to use the send value functionality, then I think it's not a problem and it is logical to have the ability to pass value.

But I don't think it is coincidental. As I understand, the system makes calls from the zero address, so if the system were ever to make use of the send value functionality it would be sending celo from the zero address to another address? Looking at the zero address on the celo explorer I can see that the balance is and always has been 0. So it seems to me that unless we undertake some systemic change there will never be a situation where the system account wants to send value.

I'm confused about your point that the vm does do value transfers using `state.SetBalance()` given that the system has a zero balance?  But we can see it is not doing that through the EVMRunner otherwise there would be a call with a non zero value.

I feel that EVMRunner is sitting on the fence, on the one hand it has elements that mean that it is only suitable for system calls but on the other it still has elements that support non system calls. I don't think this is a good state to be in, it's confusing for people and opens up avenues for misuse (I.E actually passing a non zero value in any of the places where it is currently used). So I think we should decide is it a tool just for use with system calls or not?

Reply:
well... EVMRunner is not necessarily a System EVMRunner, meaning that sender doesn't always need to be the system. That was my original idea.. .but then i was "suggested" to make that change, so that it's always the system.

So, now, that logic lives only on the `NewEVMRunner()` code, and could be easily changed by adding a parameter.

I just think you're being "too picky" about it... yet again my dear piers.... having all these conversations about what the EVMRunner should or should not be... it is exhausting.